File: leetcode/RandomizedSet.py
import random


class RandomizedSet:
    """ 
    Time complexity must be O(1) for insert, remove and getRandom
    """

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.data = []
        self.data_dict = {}
        

    def insert(self, val: int) -> list|dict:
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        """
        if val in self.data_dict:
            return False
        self.data.append(val)
        self.data_dict[val] = len(self.data) - 1
        return self.data, self.data_dict
        

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        """
        if val not in self.data_dict:
            return False
        idx = self.data_dict[val] # get the index of the value to be removed
        last_val = self.data[-1] # get the last value in the list
        self.data[idx] = last_val # replace the value to be removed with the last value
        self.data_dict[last_val] = idx # update the index of the last value
        self.data.pop() # remove the last value
        del self.data_dict[val] # remove the value from the dictionary
        return self.data, self.data_dict
        

    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        return random.choice(self.data)
    

if __name__ == "__main__":
    randomizedSet = RandomizedSet()
    print(randomizedSet.insert(1))
    print(randomizedSet.insert(3))
    print(randomizedSet.insert(2))
    print(randomizedSet.remove(3))
    print(randomizedSet.getRandom())

File: leetcode/RandomizedSet.cpp
using namespace std;
#include <stdlib.h> // for rand()
#include <time.h>
#include <vector>
#include <unordered_map>


class RandomizedSet {
public:
    /** Initialize your data structure here. */
    RandomizedSet() {
        srand(time(NULL)); // seed the random number generator
    }
    
    vector<int> v;
    unordered_map<int, int> m;
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        auto it = m.find(val);
        if(it != m.end()) return false;
        v.push_back(val);
        m[val] = v.size()-1;
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        auto it = m.find(val);
        if(it == m.end()) return false;
        int idx = it->second; // index of val in v
        m[v.back()] = idx; // v.back() is the last element in v
        swap(v[idx], v.back());
        v.pop_back();
        m.erase(it);
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        return v[rand()%v.size()]; // using % to get a random index in v
    }
};

int main() {
    RandomizedSet* obj = new RandomizedSet(); // create a new object
    bool param_1 = obj->insert(1); // insert 1
    bool param_2 = obj->remove(2); // remove 2
    int param_3 = obj->getRandom(); // get a random element
    return 0; // return 0
}